logstash-plugins/logstash-integration-jdbc
# Add another flexible way for JDBC paging (manual mode)

## Release notes

Added `jdbc_paging_mode` option to choose if use `explicit` pagination in statements and avoid the initial count query or use `auto` to delegate to the underlying library.

## What does this PR do?

This commit introduces a new configuration option to avoid the initial count select statement executed by Sequel in case of paginated queries. 
The paginated queries could be done now with initial count or not, in case `jdbc_paging_mode` is values `explicit` the plugin executes paged queries till it reach a page with less rows than the expected, instead of relying on the total row count.
In this case the SQL statement has to explicitly use the pagination keywords (like `LIMIT` and `OFFSET`) receiving as `:offset` and `:size` as implicit parameters.
When the paging mode `jdbc_paging_mode` is set to `auto` then the pagination happens automatically without the intervention of the user to create paginated query.


## Why is it important/What is the impact to the user?

In some circumstances, like stored procedure call with pagination bounds, the initial count query is simply not meaningful.
In cases where the pages to retrieve are few the initial count query could be an useful overhead.
In some complex nested queries the count could generate extra work that's not usefull

## Checklist


## Author's Checklist


- [x] check it with a local Logstash test.

## How to test this PR locally

- spin up a local DB
- configure a JDBC input query with pagination and with the new `jdbc_paging_avoid_count` and check it retrieves all the expected rows. In logs should be present a line for each paged query

```
input {
  jdbc {
    jdbc_driver_library => "/path/mysql-connector-java-8.0.26.jar"
    jdbc_driver_class => "Java::com.mysql.cj.jdbc.Driver"
    jdbc_connection_string => "jdbc:mysql://localhost:3306/test_logstash"

    jdbc_user => "user"
    jdbc_password => "s3cret"

    jdbc_default_timezone => "UTC"

    schedule => "* * * * *"

    use_column_value => true
    tracking_column => "log_id"
    tracking_column_type => "numeric"
    last_run_metadata_path => ".last_run"
    
    statement => "SELECT * FROM data_log WHERE log_id > :sql_last_value LIMIT :size OFFSET :offset"
    jdbc_paging_enabled => true
    jdbc_page_size => 7
    jdbc_paging_avoid_count => true
  }
}

output {
  stdout {
    codec => "rubydebug"
  }
}
```

## Related issues


- Superseeds #90

## Use cases


### Simple query with small result sets
In a simple query like:
```sql
SELECT * from test_table
```
the automatic pagination would generate the count query plus the paginated queries.
With small resultsets it issues 2 queries instead of one query.

### A stored procedure that requires pagination information. Suppose you have a stored like
 ```sql
CALL fetch_my_data(:sql_last_value, <offset>, <size>)
```
the automatic pagination would generate an invalid SQL statement:
```sql
SELECT count(*) AS count FROM ( CALL fetch_my_data(:sql_last_value) ) AS t1 LIMIT 1
SELECT * FROM ( CALL fetch_my_data(:sql_last_value) ) AS t1 LIMIT 10 OFFSET 0
.
.
```
which is not what's expected.

### Nested queries with pagination on the inner query
Suppose the query you want to paginate is the inner on, like in:
```sql
SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table LIMIT <offset>, <size>)
```

Without this feature, the automatic pagination would generate the following SQL:
```sql
SELECT count(*) AS count FROM ( 
  SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table) 
) AS t1 LIMIT 1;

SELECT * FROM ( 
  SELECT * FROM test_table2 WHERE id IN (SELECT * FROM test_table) 
) AS t1 LIMIT 10 OFFSET 0;
.
.
```
which is the pagination applied to the outer query instead of the inner one.
